src/run.rs

```diff
@@ -97,30 +97,44 @@ where
     <T::Service as HttpService>::Future: Send,
 {
     let listener = TcpListener::bind(addr)?;
-    let http = Arc::new(Http::new());
 
-    tokio::run({
-        listener
-            .incoming()
-            .map_err(|e| println!("failed to accept socket; err = {:?}", e))
-            .for_each(move |socket| {
-                let h = http.clone();
-
-                tokio::spawn({
-                    new_service.new_http_service()
-                        .map_err(|_| unimplemented!())
-                        .and_then(move |service| {
-                            let service = Lift::new(service);
-
-                            h.serve_connection(socket, service)
-                                .map(|_| ())
-                                .map_err(|e| {
-                                    println!("failed to serve connection; err={:?}", e);
-                                })
-                        })
-                })
-            })
-    });
+    tokio::run(serve(listener.incoming(), new_service));
 
     Ok(())
 }
+
+/// Returns a future that must be polled to process the incoming connections.
+///
+/// A non-blocking version of `run`.
+pub fn serve<S, T>(incoming: S, new_service: T) -> impl Future<Item = (), Error = ()>
+where
+    S: Stream<Item = TcpStream, Error = io::Error>,
```

Comment:
Maybe this should be switched to be a generic some trait, this would allow swapping out a real `TcpStream` for mocked out connections.

This could be `AsyncRead + AsyncWrite` (see [`serve_connection`](https://docs.rs/hyper/0.12.8/hyper/server/conn/struct.Http.html#method.serve_connection)), but I'm also wondering if we want to future proof it a bit.

Specifically, we probably want a way to get the `peer_addr` from the connection. Maybe we should add:

```rust
trait Connection: AsyncRead + AsyncWrite {
    fn peer_addr(&self) -> Result<SocketAddr> {
        // maybe a default impl returning an error?
    }
}
```

Thoughts @seanmonstar

Reply:
> to get the peer_addr from the connection

An *optional* address, right? (#71)

Reply:
Yeah, it could be `io::Result<Option<SocketAddr>>`

IMO it is a toss up if it should return `Err` or `None` :shrug:

Reply:
Hmm... Judging from how the error handling is performed in general in the library right now (`.map_err(|e| print...`) I'd say having a method that returns `Option<SocketAddr>` should suffice :)

So how about I'll go with
```rust
trait Connection: AsyncRead + AsyncWrite {
  fn peer_addr(&self) -> Option<SocketAddr> {
    None
  }
}
```
?

Reply:
@carllerche what do you think?